File: app/src/main/java/com/simplemobiletools/flashlight/helpers/MarshmallowCamera.java
package com.simplemobiletools.flashlight.helpers;

import android.annotation.TargetApi;
import android.content.Context;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraManager;
import android.os.Build;
import android.os.Handler;
import android.util.Log;

import com.simplemobiletools.flashlight.models.Events;
import com.squareup.otto.Bus;

class MarshmallowCamera {
    private static final String TAG = MyCameraImpl.class.getSimpleName();

    private CameraManager manager;
    private String cameraId;
    private Context mContext;

    @TargetApi(Build.VERSION_CODES.M)
    MarshmallowCamera(Context context) {
        mContext = context;
        manager = (CameraManager) context.getSystemService(Context.CAMERA_SERVICE);
        try {
            final String[] list = manager.getCameraIdList();
            cameraId = list[0];
        } catch (CameraAccessException ignored) {
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    void toggleMarshmallowFlashlight(final Bus bus, boolean enable) {
        try {
            manager.setTorchMode(cameraId, enable);
        } catch (CameraAccessException e) {
            Log.e(TAG, "toggle marshmallow flashlight " + e.getMessage());

            Runnable mainRunnable = new Runnable() {
                @Override
                public void run() {
                    bus.post(new Events.CameraUnavailable());
                }
            };
            new Handler(mContext.getMainLooper()).post(mainRunnable);
        }
    }
}

File: app/src/main/kotlin/com/simplemobiletools/flashlight/helpers/MarshmallowCamera.kt
package com.simplemobiletools.flashlight.helpers

import android.annotation.TargetApi
import android.content.Context
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.Handler
import com.simplemobiletools.commons.extensions.showErrorToast
import com.simplemobiletools.flashlight.models.Events
import org.greenrobot.eventbus.EventBus

internal class MarshmallowCamera constructor(val context: Context) {

    private val manager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
    private var cameraId: String? = null

    init {
        try {
            cameraId = manager.cameraIdList[0] ?: "0"
        } catch (e: Exception) {
            context.showErrorToast(e)
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    fun toggleMarshmallowFlashlight(enable: Boolean) {
        try {
            manager.setTorchMode(cameraId!!, enable)
        } catch (e: Exception) {
            context.showErrorToast(e)
            val mainRunnable = Runnable {
                EventBus.getDefault().post(Events.CameraUnavailable())
            }
            Handler(context.mainLooper).post(mainRunnable)
        }
    }
}
